### PR TITLE
Add the ability to pass backend specific options through CompilerConfig.

### PR DESCRIPTION
**Pull request by nlewycky**
# Description
Add the ability to pass backend specific options through CompilerConfig.

Use this to replace wasmer_llvm_backend::GLOBAL_OPTIONS.

This adds a new item to CompilerConfig which can be any struct defined by a backend for backend specific configuration options. So far, we only implement the options that were previously in GLOBAL_OPTIONS, these move to wasmer.rs. The plan is to use these same callbacks for testing in llvm-backend-tests.

# Review

- [ ] Add a short description of the the change to the CHANGELOG.md file

### CODE REVIEW

File: lib/llvm-backend/src/backend.rs [hunk 1]
```diff
@@ -176,23 +175,22 @@ impl LLVMBackend {
         _stackmaps: &StackmapRegistry,
         _module_info: &ModuleInfo,
         target_machine: &TargetMachine,
+        llvm_callbacks: &Option<Rc<RefCell<dyn LLVMCallbacks>>>,
```
Comment (MarkMcCaskey) [inline lib/llvm-backend/src/backend.rs]: Might make sense to make this take a &mut directly, the fact that it's an `Rc<RefCell<>>` doesn't seem relevant to the function

File: lib/runtime-core/src/backend.rs [hunk 2]
```diff
@@ -200,6 +200,15 @@ pub struct Features {
     pub threads: bool,
 }
 
+#[derive(Debug)]
+pub struct BackendCompilerConfig(pub Box<dyn Any + 'static>);
```
Comment (MarkMcCaskey) [inline lib/runtime-core/src/backend.rs]: Some doc comments and maybe some implementation-specific comments would be good to have here -- this code is a bit abstract and can be scary

File: src/bin/wasmer.rs [hunk 3]
```diff
@@ -558,25 +595,22 @@ fn execute_wasm(options: &Run) -> Result<(), String> {
...
+            backend_specific_config = Some(BackendCompilerConfig(Box::new(LLVMBackendConfig {
+                callbacks: Some(Rc::new(RefCell::new(options.backend_llvm_options.clone()))),
+            })))
         }
     }
 
+    let compiler: Box<dyn Compiler> = match get_compiler_by_backend(options.backend, options) {
+        Some(x) => x,
+        None => return Err("the requested backend is not enabled".into()),
+    };
```
Comment (MarkMcCaskey) [inline src/bin/wasmer.rs]: ⛳️ 
```
let compiler: Box<dyn Compiler> =
     get_compiler_by_backend(options.backend, options).ok_or_else(|| "the requested backend is not enabled".into())?;
```

File: lib/runtime-core/src/backend.rs (hunk 2 above)
Comment (nlewycky) [reply lib/runtime-core/src/backend.rs]: I added some, let me know if you think I got it or if there's more to add / something I forgot.

File: lib/llvm-backend/src/backend.rs (hunk 1 above)
Comment (MarkMcCaskey) [reply lib/llvm-backend/src/backend.rs]: You should be able to have an `Option<&mut dyn LLVMCallbacks>` I believe

Comment (MarkMcCaskey) [reply lib/llvm-backend/src/backend.rs]: Something like `val.map(|inner| inner.borrow_mut())` might do the trick

Comment (nlewycky) [reply lib/llvm-backend/src/backend.rs]: Just marking down for review that Mark and I paired on this for a while and weren't able to make it work.

Comment (MarkMcCaskey) [reply lib/llvm-backend/src/backend.rs]: Or that the cure seemed worse than the disease, so we chose to not bother
